python/lsst/pipe/base/graph/graph.py

```diff
@@ -0,0 +1,779 @@
+    def graph(self) -> nx.DiGraph:
+        """Returns a graph representing the relations between all the
+        `QuantumNode`s. Larely it should be preferred to iterate over, and use
+        methods of this class, but sometimes direct access to the networkx
+        object may be helpful
+
+        Returns
+        -------
+        graph : `networkx.Digraph`
+            Internal datastructure that holds relations of `QuantumNode`s
+        """
+        return self._connectedQuanta
+
+    def taskDefForNode(self, node: QuantumNode) -> TaskDef:
+        """Retrives the TaskDef associated with a `QuantumNode`.
+
+        Parameters
+        ----------
+        node : `QuantumNode`
+            Node for which the associated `TaskDef` is to be found
+
+        Returns
+        -------
+        `TaskDef`
+            The requested `TaskDef`
+        """
+        return self._quantumToTask[node.nodeNumber]
+
+    def getQuantumNodeByNodeNumber(self, nodeNumber: int) -> QuantumNode:
+        """Lookup a `QuantumNode` from a number associated with the node.
+
+        Parameters
+        ----------
+        nodeNumber : `int`
+            The number associated with a node
+
+        Returns
+        -------
+        `QuantumNode`
+            The node corresponding with input number
+
+        Raises
+        ------
+        `IndexError`
+            If the requested nodeNumber is greater than the number of nodes in
+            the graph.
+        """
+        return self._nodeList[nodeNumber]
+
+    @property
+    def inputQuanta(self) -> List[QuantumNode]:
+        """Makes a `list` of all `QuantumNode`s that are 'input' nodes to the
+        graph, meaning those nodes to not depend on any other nodes in the
+        graph.
+
+        Returns
+        -------
+        inputNodes : list of `QuantumNode`
+        """
+        return [q for q, n in self._connectedQuanta.in_degree if n == 0]
+
+    @property
+    def outputQuanta(self) -> List[QuantumNode]:
+        """Makes a `list` of all `QuantumNode`s that are 'output' nodes to the
+        graph, meaning those nodes have no nodes that depend them in the graph.
+
+        Returns
+        -------
+        outputNodes : list of `QuantumNode`
+        """
+        return [q for q, n in self._connectedQuanta.out_degree if n == 0]
+
+    def countQuanta(self) -> int:
+        """Returns the number of nodes in the graph
+
+        Returns
+        -------
+        int
+            Number of nodes in the graph
+        """
+        if self._count is None:
+            self._count = int(sum(len(values) for values in self._quanta.values()))
+        return self._count
+
+    def quantaForTask(self, taskDef: TaskDef) -> Set[Quantum]:
+        """Returns all the `Quantum` associated with a `TaskDef`
+
+        Parameters
+        ----------
+        taskDef : `TaskDef`
+            The `TaskDef` for which `Quantum` are to be queried
+
+        Returns
+        -------
+        set of `Quantum`
```

Comment:
I thought return values needed to be named, just like parameters.  I may be wrong, but please check that this renders as you expect.  Also, `set` should have backticks, as should `list` in many other docstrings here.

Reply:
yes, they need to be named.

Reply:
Sorry I was used to numpy-docs where they are not required and either forgot or did not know that we had a different standard.